analysis/study_definition.py

```diff
@@ -1,33 +1,191 @@
+        returning="date",
+        find_first_match_in_period=True,
+        include_month=True,
+        return_expectations={
+            "incidence": 0.2,
+            "date": {"earliest": "2010-01-01", "latest": "2020-02-29"},
+        },
+    )
+
+def medication_latest(med_codelist):
+    return patients.with_these_medications(
+        med_codelist,
+        between=["2010-01-01", "2020-02-29"],
+        returning="date",
+        find_last_match_in_period=True,
+        include_month=True,
+        return_expectations={
+            "incidence": 0.2,
+            "date": {"earliest": "2010-01-01", "latest": "2020-02-29"},
+        },
+    )
+
+def medication_counts_and_dates(var_name, med_codelist_file):
+    """
+    Generates dictionary of covariats for a medication including counts and dates
+    
+    Takes a variable prefix and medication codelist filename (minus .csv)
+    Returns a dictionary suitable for unpacking into the main study definition
+    This will include all five of the items defined in the functions above
+    """
+    
+    definitions={}
+    if (med_codelist_file[0:5] == "cross"):
+        med_codelist_file = "crossimid-codelists/" + med_codelist_file
+    else:
+        med_codelist_file = "codelists/" + med_codelist_file
+    med_codelist=codelist_from_csv(med_codelist_file + ".csv", system="snomed", column="snomed_id")
+    med_functions=[
+        ("3m_0m", medication_count_3m_0m),
+        ("6m_3m", medication_count_6m_3m),
+        ("12m_6m", medication_count_12m_6m),
+        ("earliest", medication_earliest),
+        ("latest", medication_latest)
+    ]
+    for (suffix, fun) in med_functions:
+        definitions[var_name + "_" + suffix] = fun(med_codelist)
+    return definitions
+
+def medication_counts_and_dates_all(meds_list):
+    """
+    Generate dictionary of covariates for list of medications including counts and dates
+    
+    Takes a list of tuples of the form (variable prefix, medication codelist filename (minus .csv))
+    Returns a dictionary suitable for unpacking into the main study definition
+    For each tuple, this will include all of the items specified in `medication_counts_and_dates`
+    """
+    definitions={}
+    for (var_name, med_codelist_file) in meds_list:
+        definitions.update(medication_counts_and_dates(var_name, med_codelist_file))
+    return definitions
 
 study = StudyDefinition(
     # Configure the expectations framework
     default_expectations={
         "date": {"earliest": "1900-01-01", "latest": "today"},
-        "rate": "exponential_increase",
+        "rate": "uniform",
+        "incidence": 0.1,
     },
     # This line defines the study population
     population=patients.registered_with_one_practice_between(
-        "2019-02-01", "2020-02-01"
+        "2019-03-01", "2020-03-01"
+    ),
+    # Outcomes
+    icu_date_admitted=patients.admitted_to_icu(
+        on_or_after="2020-03-01",
+        include_day=True,
```

Comment:
we generally don't include the date unless compelling reason for data minimisation principle

Reply:
We do need the date for the STATA analysis, though it won't be exposed at the end of the STATA script to export